### PR TITLE
Include feature test to verify basic high-availability workflow 

### LINKED ISSUE #531: Create a sample test workflow with associated playbooks to perform a fault injection test
## Is this a BUG REPORT or FEATURE REQUEST?
 FEATURE REQUEST

The current test jobs that are run along with application pods in the integration tests (such as the vdbench & fio bench/sql-loadgen/tpcc-client etc., ) need to include the ability to trigger and revert faults such as network faults/container kill/node panic etc..,

This will be especially helpful for HA-related integration tests.

This issue tracks the creation of a pilot/sample fault-injection test workflow on openebs storage in kubernetes cluster using :

 - Ansible playbooks which can perform a simple "fault-injection" 

This will help us with identifying the pattern/template for such tests around checking the following: 

- Desired system state/pre-requisite
- Fault-injection step
- Expected result
- Recovery post failure
- Cleanup 

Once this workflow is solidified, we can create test containers which can be used in kubernetes jobs to perform these tests - with scope for integration of different fault-injection frameworks.

### PR DESCRIPTION
**Pull request by ksatchit**
**What this PR does / why we need it**:

HA is a critical feature of the OpenEBS storage solution. This commit adds the ansible playbook and task files necessary to execute a basic ha testcase that causes the ctrl pod reschedule from one node to another

**Which issue this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*: fixes #531  

**Special notes for your reviewer**:

2. How does this change address the issue?

Typically, fault-injection is used to test resiliency (here in the context of HA). Fault-injection tests generally include the following steps:

- Test bed/env creation
- Pre-fault health checks/prepare for fault-injection
- Fault injection
- Verify system's fault-tolerance/Recovery capabilities
- Post recovery health-checks
- Revert fault
- Cleanup

The included test playbook attempts to create a "HA/Resiliency test template" that can be followed in the future while building test playbooks in future. It uses "kubectl cordon" and "kubectl delete pod" commands to modify the infrastructure, inject fault and generally simulate HA.

The application used is percona, which is checked for "Running" status and "zero" restarts during HA as part of verifying the tolerance and health-check. Also, some sample db query is performed after ctrl pod reschedule.

Files in this PR include : 

- ha-vars.yml - Holds test variables
- ha-prerequisites.yml - Set of tasks to be performed before starting test (packages etc.,)
- ha-cleanup.yml - Cleanup routines
- ha.yml - Main test playbook (above three are statically included into this yaml)

**Note**: Under ideal circumstances we can have live-load on the db while performing this test, something that is ideally achievable by the ansible "async" module. However, due  to some open issues in it, we have not included it in the playbook. Instead, we have included logic to check status/restart count of pod and perform db writes post ha to check sanity. 

Resiliency tests in future are expected to make use of tools like [pumba](https://github.com/gaia-adm/pumba) and stronger verification procedures.

### CODE REVIEW

**Comment by ksatchit** [comment]
Here is the output of the playbook run with verbose enabled

```
ciuser@OpenEBSClient:~/openebs/e2e/ansible$ ansible-playbook playbooks/feature/ha/ha.yml -vv
Using /home/ciuser/openebs/e2e/ansible/ansible.cfg as config file
statically included: /home/ciuser/openebs/e2e/ansible/playbooks/feature/ha/ha-prerequisites.yml
statically included: /home/ciuser/openebs/e2e/ansible/playbooks/feature/ha/ha-cleanup.yml

PLAYBOOK: ha.yml *******************************************************************************************************
1 plays in playbooks/feature/ha/ha.yml

PLAY [localhost] *******************************************************************************************************

TASK [Gathering Facts] *************************************************************************************************
ok: [localhost]
META: ran handlers

TASK [Run apt-get update via shell] ************************************************************************************
task path: /home/ciuser/openebs/e2e/ansible/playbooks/feature/ha/ha-prerequisites.yml:2
 [WARNING]: Consider using apt module rather than running apt-get

changed: [localhost -> None] => {"changed": true, "cmd": "apt-get update", "delta": "0:00:18.546668", "end": "2017-11-07 12:39:51.864918", "rc": 0, "start": "2017-11-07 12:39:33.318250", "stderr": "", "stderr_lines": [], "stdout": "Get:1 http://security.ubuntu.com/ubuntu xenial-security InRelease [102 kB]\nHit:2 http://us.archive.ubuntu.com/ubuntu xenial InRelease\nGet:4 http://security.ubuntu.com/ubuntu xenial-security/main amd64 Packages [381 kB]\nGet:5 http://us.archive.ubuntu.com/ubuntu xenial-updates InRelease [102 kB]\nGet:6 http://security.ubuntu.com/ubuntu xenial-security/main i386 Packages [352 kB]\nGet:7 http://us.archive.ubuntu.com/ubuntu xenial-backports InRelease [102 kB]\nGet:8 http://us.archive.ubuntu.com/ubuntu xenial-updates/main amd64 Packages [652 kB]\nGet:9 http://us.archive.ubuntu.com/ubuntu xenial-updates/main i386 Packages [616 kB]\nGet:10 http://us.archive.ubuntu.com/ubuntu xenial-updates/universe amd64 Packages [543 kB]\nHit:3 https://packages.cloud.google.com/apt kubernetes-xenial InRelease\nGet:11 http://us.archive.ubuntu.com/ubuntu xenial-updates/universe i386 Packages [517 kB]\nFetched 3,368 kB in 15s (223 kB/s)\nReading package lists...", "stdout_lines": ["Get:1 http://security.ubuntu.com/ubuntu xenial-security InRelease [102 kB]", "Hit:2 http://us.archive.ubuntu.com/ubuntu xenial InRelease", "Get:4 http://security.ubuntu.com/ubuntu xenial-security/main amd64 Packages [381 kB]", "Get:5 http://us.archive.ubuntu.com/ubuntu xenial-updates InRelease [102 kB]", "Get:6 http://security.ubuntu.com/ubuntu xenial-security/main i386 Packages [352 kB]", "Get:7 http://us.archive.ubuntu.com/ubuntu xenial-backports InRelease [102 kB]", "Get:8 http://us.archive.ubuntu.com/ubuntu xenial-updates/main amd64 Packages [652 kB]", "Get:9 http://us.archive.ubuntu.com/ubuntu xenial-updates/main i386 Packages [616 kB]", "Get:10 http://us.archive.ubuntu.com/ubuntu xenial-updates/universe amd64 Packages [543 kB]", "Hit:3 https://packages.cloud.google.com/apt kubernetes-xenial InRelease", "Get:11 http://us.archive.ubuntu.com/ubuntu xenial-updates/universe i386 Packages [517 kB]", "Fetched 3,368 kB in 15s (223 kB/s)", "Reading package lists..."]}

TASK [Install python-pip on K8S-minion] ********************************************************************************
task path: /home/ciuser/openebs/e2e/ansible/playbooks/feature/ha/ha-prerequisites.yml:8
changed: [localhost -> None] => (item=mysql-client) => {"changed": true, "cmd": "apt-get install -y mysql-client", "delta": "0:00:11.776182", "end": "2017-11-07 12:40:04.323030", "item": "mysql-client", "rc": 0, "start": "2017-11-07 12:39:52.546848", "stderr": "", "stderr_lines": [], "stdout": "Reading package lists...\nBuilding dependency tree...\nReading state information...\nThe following packages were automatically installed and are no longer required:\n  gir1.2-glib-2.0 libdbus-glib-1-2 libgirepository-1.0-1 python-newt\nUse 'sudo apt autoremove' to remove them.\nThe following additional packages will be installed:\n  mysql-client-5.7 mysql-client-core-5.7 mysql-common\nThe following NEW packages will be installed:\n  mysql-client mysql-client-5.7 mysql-client-core-5.7 mysql-common\n0 upgraded, 4 newly installed, 0 to remove and 193 not upgraded.\nNeed to get 8,039 kB of archives.\nAfter this operation, 65.7 MB of additional disk space will be used.\nGet:1 http://us.archive.ubuntu.com/ubuntu xenial-updates/main amd64 mysql-client-core-5.7 amd64 5.7.20-0ubuntu0.16.04.1 [6,339 kB]\nGet:2 http://us.archive.ubuntu.com/ubuntu xenial-updates/main amd64 mysql-common all 5.7.20-0ubuntu0.16.04.1 [15.6 kB]\nGet:3 http://us.archive.ubuntu.com/ubuntu xenial-updates/main amd64 mysql-client-5.7 amd64 5.7.20-0ubuntu0.16.04.1 [1,675 kB]\nGet:4 http://us.archive.ubuntu.com/ubuntu xenial-updates/main amd64 mysql-client all 5.7.20-0ubuntu0.16.04.1 [10.1 kB]\nFetched 8,039 kB in 6s (1,201 kB/s)\nSelecting previously unselected package mysql-client-core-5.7.\r\n(Reading database ... \r(Reading database ... 5%\r(Reading database ... 10%\r(Reading database ... 15%\r(Reading database ... 20%\r(Reading database ... 25%\r(Reading database ... 30%\r(Reading database ... 35%\r(Reading database ... 40%\r(Reading database ... 45%\r(Reading database ... 50%\r(Reading database ... 55%\r(Reading database ... 60%\r(Reading database ... 65%\r(Reading database ... 70%\r(Reading database ... 75%\r(Reading database ... 80%\r(Reading database ... 85%\r(Reading database ... 90%\r(Reading database ... 95%\r(Reading database ... 100%\r(Reading database ... 92870 files and directories currently installed.)\r\nPreparing to unpack .../mysql-client-core-5.7_5.7.20-0ubuntu0.16.04.1_amd64.deb ...\r\nUnpacking mysql-client-core-5.7 (5.7.20-0ubuntu0.16.04.1) ...\r\nSelecting previously unselected package mysql-common.\r\nPreparing to unpack .../mysql-common_5.7.20-0ubuntu0.16.04.1_all.deb ...\r\nUnpacking mysql-common (5.7.20-0ubuntu0.16.04.1) ...\r\nSelecting previously unselected package mysql-client-5.7.\r\nPreparing to unpack .../mysql-client-5.7_5.7.20-0ubuntu0.16.04.1_amd64.deb ...\r\nUnpacking mysql-client-5.7 (5.7.20-0ubuntu0.16.04.1) ...\r\nSelecting previously unselected package mysql-client.\r\nPreparing to unpack .../mysql-client_5.7.20-0ubuntu0.16.04.1_all.deb ...\r\nUnpacking mysql-client (5.7.20-0ubuntu0.16.04.1) ...\r\nProcessing triggers for man-db (2.7.5-1) ...\r\nSetting up mysql-client-core-5.7 (5.7.20-0ubuntu0.16.04.1) ...\r\nSetting up mysql-common (5.7.20-0ubuntu0.16.04.1) ...\r\nupdate-alternatives: using /etc/mysql/my.cnf.fallback to provide /etc/mysql/my.cnf (my.cnf) in auto mode\r\nSetting up mysql-client-5.7 (5.7.20-0ubuntu0.16.04.1) ...\r\nSetting up mysql-client (5.7.20-0ubuntu0.16.04.1) ...", "stdout_lines": ["Reading package lists...", "Building dependency tree...", "Reading state information...", "The following packages were automatically installed and are no longer required:", "  gir1.2-glib-2.0 libdbus-glib-1-2 libgirepository-1.0-1 python-newt", "Use 'sudo apt autoremove' to remove them.", "The following additional packages will be installed:", "  mysql-client-5.7 mysql-client-core-5.7 mysql-common", "The following NEW packages will be installed:", "  mysql-client mysql-client-5.7 mysql-client-core-5.7 mysql-common", "0 upgraded, 4 newly installed, 0 to remove and 193 not upgraded.", "Need to get 8,039 kB of archives.", "After this operation, 65.7 MB of additional disk space will be used.", "Get:1 http://us.archive.ubuntu.com/ubuntu xenial-updates/main amd64 mysql-client-core-5.7 amd64 5.7.20-0ubuntu0.16.04.1 [6,339 kB]", "Get:2 http://us.archive.ubuntu.com/ubuntu xenial-updates/main amd64 mysql-common all 5.7.20-0ubuntu0.16.04.1 [15.6 kB]", "Get:3 http://us.archive.ubuntu.com/ubuntu xenial-updates/main amd64 mysql-client-5.7 amd64 5.7.20-0ubuntu0.16.04.1 [1,675 kB]", "Get:4 http://us.archive.ubuntu.com/ubuntu xenial-updates/main amd64 mysql-client all 5.7.20-0ubuntu0.16.04.1 [10.1 kB]", "Fetched 8,039 kB in 6s (1,201 kB/s)", "Selecting previously unselected package mysql-client-core-5.7.", "(Reading database ... ", "(Reading database ... 5%", "(Reading database ... 10%", "(Reading database ... 15%", "(Reading database ... 20%", "(Reading database ... 25%", "(Reading database ... 30%", "(Reading database ... 35%", "(Reading database ... 40%", "(Reading database ... 45%", "(Reading database ... 50%", "(Reading database ... 55%", "(Reading database ... 60%", "(Reading database ... 65%", "(Reading database ... 70%", "(Reading database ... 75%", "(Reading database ... 80%", "(Reading database ... 85%", "(Reading database ... 90%", "(Reading database ... 95%", "(Reading database ... 100%", "(Reading database ... 92870 files and directories currently installed.)", "Preparing to unpack .../mysql-client-core-5.7_5.7.20-0ubuntu0.16.04.1_amd64.deb ...", "Unpacking mysql-client-core-5.7 (5.7.20-0ubuntu0.16.04.1) ...", "Selecting previously unselected package mysql-common.", "Preparing to unpack .../mysql-common_5.7.20-0ubuntu0.16.04.1_all.deb ...", "Unpacking mysql-common (5.7.20-0ubuntu0.16.04.1) ...", "Selecting previously unselected package mysql-client-5.7.", "Preparing to unpack .../mysql-client-5.7_5.7.20-0ubuntu0.16.04.1_amd64.deb ...", "Unpacking mysql-client-5.7 (5.7.20-0ubuntu0.16.04.1) ...", "Selecting previously unselected package mysql-client.", "Preparing to unpack .../mysql-client_5.7.20-0ubuntu0.16.04.1_all.deb ...", "Unpacking mysql-client (5.7.20-0ubuntu0.16.04.1) ...", "Processing triggers for man-db (2.7.5-1) ...", "Setting up mysql-client-core-5.7 (5.7.20-0ubuntu0.16.04.1) ...", "Setting up mysql-common (5.7.20-0ubuntu0.16.04.1) ...", "update-alternatives: using /etc/mysql/my.cnf.fallback to provide /etc/mysql/my.cnf (my.cnf) in auto mode", "Setting up mysql-client-5.7 (5.7.20-0ubuntu0.16.04.1) ...", "Setting up mysql-client (5.7.20-0ubuntu0.16.04.1) ..."], "warnings": ["Consider using apt module rather than running apt-get"]}

TASK [Get $HOME of K8s master for kubernetes user] *********************************************************************
task path: /home/ciuser/openebs/e2e/ansible/playbooks/feature/ha/ha.yml:12
changed: [localhost -> None] => {"changed": true, "cmd": "source ~/.profile; echo $HOME", "delta": "0:00:00.003372", "end": "2017-11-07 12:40:04.715635", "rc": 0, "start": "2017-11-07 12:40:04.712263", "stderr": "", "stderr_lines": [], "stdout": "/home/karthik", "stdout_lines": ["/home/karthik"]}

TASK [Download YAML for percona mysql plugin] **************************************************************************
task path: /home/ciuser/openebs/e2e/ansible/playbooks/feature/ha/ha.yml:19
changed: [localhost -> None] => {"attempts": 1, "changed": true, "checksum_dest": null, "checksum_src": "30856d328bde5dfdf9e3026c94f7bef160949ba5", "dest": "/home/karthik/demo-percona-mysql-pvc.yaml", "gid": 1000, "group": "karthik", "md5sum": "9a23f470626221c60d8b920af95261ca", "mode": "0664", "msg": "OK (737 bytes)", "owner": "karthik", "size": 737, "src": "/tmp/tmpNa_Uog", "state": "file", "status_code": 200, "uid": 1000, "url": "https://raw.githubusercontent.com/openebs/openebs/master/k8s/demo/percona/demo-percona-mysql-pvc.yaml"}

TASK [Check whether maya-apiserver pod is deployed] ********************************************************************
task path: /home/ciuser/openebs/e2e/ansible/playbooks/feature/ha/ha.yml:30
changed: [localhost -> None] => {"attempts": 1, "changed": true, "cmd": "source ~/.profile; kubectl get pods -l name=maya-apiserver --no-headers", "delta": "0:00:01.372366", "end": "2017-11-07 12:40:14.410565", "rc": 0, "start": "2017-11-07 12:40:13.038199", "stderr": "", "stderr_lines": [], "stdout": "maya-apiserver-3416621614-g6tmq   1/1       Running   0         4d", "stdout_lines": ["maya-apiserver-3416621614-g6tmq   1/1       Running   0         4d"]}

TASK [Deploy percona mysql pod] ****************************************************************************************
task path: /home/ciuser/openebs/e2e/ansible/playbooks/feature/ha/ha.yml:42
changed: [localhost -> None] => {"changed": true, "cmd": "source ~/.profile; kubectl create -f demo-percona-mysql-pvc.yaml", "delta": "0:00:00.642361", "end": "2017-11-07 12:40:15.409911", "rc": 0, "start": "2017-11-07 12:40:14.767550", "stderr": "", "stderr_lines": [], "stdout": "pod \"percona\" created\npersistentvolumeclaim \"demo-vol1-claim\" created", "stdout_lines": ["pod \"percona\" created", "persistentvolumeclaim \"demo-vol1-claim\" created"]}

TASK [Confirm pod status is running] ***********************************************************************************
task path: /home/ciuser/openebs/e2e/ansible/playbooks/feature/ha/ha.yml:48
FAILED - RETRYING: Confirm pod status is running (15 retries left).
FAILED - RETRYING: Confirm pod status is running (14 retries left).
changed: [localhost -> None] => {"attempts": 3, "changed": true, "cmd": "source ~/.profile; kubectl get pods -l name=percona --no-headers", "delta": "0:00:07.037959", "end": "2017-11-07 12:44:30.120198", "rc": 0, "start": "2017-11-07 12:44:23.082239", "stderr": "", "stderr_lines": [], "stdout": "percona   1/1       Running   0         4m", "stdout_lines": ["percona   1/1       Running   0         4m"]}

TASK [Get storage ctrl pod name] ***************************************************************************************
task path: /home/ciuser/openebs/e2e/ansible/playbooks/feature/ha/ha.yml:60
changed: [localhost -> None] => {"changed": true, "cmd": "source ~/.profile; kubectl get pods -l openebs/controller=jiva-controller --no-headers", "delta": "0:00:00.498554", "end": "2017-11-07 12:44:31.014326", "rc": 0, "start": "2017-11-07 12:44:30.515772", "stderr": "", "stderr_lines": [], "stdout": "pvc-b42c2d3a-c38a-11e7-9be8-000c298ff5fc-ctrl-1983520977-l7r2m   1/1       Running   0         4m", "stdout_lines": ["pvc-b42c2d3a-c38a-11e7-9be8-000c298ff5fc-ctrl-1983520977-l7r2m   1/1       Running   0         4m"]}

TASK [Set ctrl pod name to variable] ***********************************************************************************
task path: /home/ciuser/openebs/e2e/ansible/playbooks/feature/ha/ha.yml:70
ok: [localhost] => {"ansible_facts": {"ctrl_pod_name": "pvc-b42c2d3a-c38a-11e7-9be8-000c298ff5fc-ctrl-1983520977-l7r2m"}, "changed": false}

TASK [Get node on which the pod is scheduled] **************************************************************************
task path: /home/ciuser/openebs/e2e/ansible/playbooks/feature/ha/ha.yml:74
changed: [localhost -> None] => {"changed": true, "cmd": "source ~/.profile; kubectl describe pod pvc-b42c2d3a-c38a-11e7-9be8-000c298ff5fc-ctrl-1983520977-l7r2m | grep -w 'Node:'", "delta": "0:00:00.345602", "end": "2017-11-07 12:44:31.837876", "rc": 0, "start": "2017-11-07 12:44:31.492274", "stderr": "", "stderr_lines": [], "stdout": "Node:\t\tmayahost01/20.10.49.12", "stdout_lines": ["Node:\t\tmayahost01/20.10.49.12"]}

TASK [Store the ctrl pod node name in a variable] **********************************************************************
task path: /home/ciuser/openebs/e2e/ansible/playbooks/feature/ha/ha.yml:83
ok: [localhost] => {"ansible_facts": {"ctrl_node": "mayahost01"}, "changed": false}

TASK [Make the current ctrl node unschedulable] ************************************************************************
task path: /home/ciuser/openebs/e2e/ansible/playbooks/feature/ha/ha.yml:87
changed: [localhost -> None] => {"changed": true, "cmd": "source ~/.profile; kubectl cordon mayahost01", "delta": "0:00:00.319193", "end": "2017-11-07 12:44:32.674365", "rc": 0, "start": "2017-11-07 12:44:32.355172", "stderr": "", "stderr_lines": [], "stdout": "node \"mayahost01\" cordoned", "stdout_lines": ["node \"mayahost01\" cordoned"]}

TASK [Confirm ctrl node is unschedulable] ******************************************************************************
task path: /home/ciuser/openebs/e2e/ansible/playbooks/feature/ha/ha.yml:94
changed: [localhost -> None] => {"changed": true, "cmd": "source ~/.profile; kubectl get nodes | grep mayahost01", "delta": "0:00:00.374601", "end": "2017-11-07 12:44:33.416954", "failed": false, "failed_when_result": false, "rc": 0, "start": "2017-11-07 12:44:33.042353", "stderr": "", "stderr_lines": [], "stdout": "mayahost01   Ready,SchedulingDisabled   10d       v1.7.7", "stdout_lines": ["mayahost01   Ready,SchedulingDisabled   10d       v1.7.7"]}

TASK [Delete the ctrl pod to force reschedule] *************************************************************************
task path: /home/ciuser/openebs/e2e/ansible/playbooks/feature/ha/ha.yml:102
changed: [localhost -> None] => {"changed": true, "cmd": "source ~/.profile; kubectl delete pod pvc-b42c2d3a-c38a-11e7-9be8-000c298ff5fc-ctrl-1983520977-l7r2m", "delta": "0:00:00.314813", "end": "2017-11-07 12:44:34.113765", "failed": false, "failed_when_result": false, "rc": 0, "start": "2017-11-07 12:44:33.798952", "stderr": "", "stderr_lines": [], "stdout": "pod \"pvc-b42c2d3a-c38a-11e7-9be8-000c298ff5fc-ctrl-1983520977-l7r2m\" deleted", "stdout_lines": ["pod \"pvc-b42c2d3a-c38a-11e7-9be8-000c298ff5fc-ctrl-1983520977-l7r2m\" deleted"]}

TASK [Check if ctrl pod is restarted within 30s] ***********************************************************************
task path: /home/ciuser/openebs/e2e/ansible/playbooks/feature/ha/ha.yml:110
FAILED - RETRYING: Check if ctrl pod is restarted within 30s (6 retries left).
changed: [localhost -> None] => {"attempts": 2, "changed": true, "cmd": "source ~/.profile; kubectl get pods -l openebs/controller=jiva-controller", "delta": "0:00:00.287289", "end": "2017-11-07 12:44:40.593672", "rc": 0, "start": "2017-11-07 12:44:40.306383", "stderr": "", "stderr_lines": [], "stdout": "NAME                                                             READY     STATUS    RESTARTS   AGE\npvc-b42c2d3a-c38a-11e7-9be8-000c298ff5fc-ctrl-1983520977-75j0q   1/1       Running   0          6s", "stdout_lines": ["NAME                                                             READY     STATUS    RESTARTS   AGE", "pvc-b42c2d3a-c38a-11e7-9be8-000c298ff5fc-ctrl-1983520977-75j0q   1/1       Running   0          6s"]}

TASK [Get percona pod details post ha] *********************************************************************************
task path: /home/ciuser/openebs/e2e/ansible/playbooks/feature/ha/ha.yml:122
changed: [localhost -> None] => {"changed": true, "cmd": "source ~/.profile; kubectl get pods -l name=percona -o wide --no-headers", "delta": "0:00:00.311529", "end": "2017-11-07 12:44:41.294618", "rc": 0, "start": "2017-11-07 12:44:40.983089", "stderr": "", "stderr_lines": [], "stdout": "percona   1/1       Running   0         4m        10.44.0.2   mayahost02", "stdout_lines": ["percona   1/1       Running   0         4m        10.44.0.2   mayahost02"]}

TASK [Set pod variables to facts] **************************************************************************************
task path: /home/ciuser/openebs/e2e/ansible/playbooks/feature/ha/ha.yml:132
ok: [localhost] => {"ansible_facts": {"pod_ip": "10.44.0.2", "pod_restarts": "0", "pod_status": "Running"}, "changed": false}

TASK [Verify percona is running w/o restarts] **************************************************************************
task path: /home/ciuser/openebs/e2e/ansible/playbooks/feature/ha/ha.yml:138
ok: [localhost] => {
    "changed": false,
    "failed": false,
    "failed_when_result": false,
    "msg": "Percona pod health-check is successful"
}

TASK [Create a test db on percona to confirm health] *******************************************************************
task path: /home/ciuser/openebs/e2e/ansible/playbooks/feature/ha/ha.yml:143
changed: [localhost -> None] => {"changed": true, "cmd": "mysql -uroot -pk8sDem0 -h 10.44.0.2 -e \"CREATE DATABASE Inventory;\"", "delta": "0:00:00.109490", "end": "2017-11-07 12:44:42.012633", "failed": false, "failed_when_result": false, "rc": 0, "start": "2017-11-07 12:44:41.903143", "stderr": "mysql: [Warning] Using a password on the command line interface can be insecure.", "stderr_lines": ["mysql: [Warning] Using a password on the command line interface can be insecure."], "stdout": "", "stdout_lines": []}

TASK [Uncordon the K8S node as part of cleanup] ************************************************************************
task path: /home/ciuser/openebs/e2e/ansible/playbooks/feature/ha/ha.yml:151
changed: [localhost -> None] => {"changed": true, "cmd": "source ~/.profile; kubectl uncordon mayahost01", "delta": "0:00:00.331000", "end": "2017-11-07 12:44:42.782120", "rc": 0, "start": "2017-11-07 12:44:42.451120", "stderr": "", "stderr_lines": [], "stdout": "node \"mayahost01\" uncordoned", "stdout_lines": ["node \"mayahost01\" uncordoned"]}

TASK [Delete percona mysql pod] ****************************************************************************************
task path: /home/ciuser/openebs/e2e/ansible/playbooks/feature/ha/ha-cleanup.yml:2
changed: [localhost -> None] => {"changed": true, "cmd": "source ~/.profile; kubectl delete -f demo-percona-mysql-pvc.yaml", "delta": "0:00:00.342231", "end": "2017-11-07 12:44:43.522594", "rc": 0, "start": "2017-11-07 12:44:43.180363", "stderr": "", "stderr_lines": [], "stdout": "pod \"percona\" deleted\npersistentvolumeclaim \"demo-vol1-claim\" deleted", "stdout_lines": ["pod \"percona\" deleted", "persistentvolumeclaim \"demo-vol1-claim\" deleted"]}

TASK [Confirm percona pod has been deleted] ****************************************************************************
task path: /home/ciuser/openebs/e2e/ansible/playbooks/feature/ha/ha-cleanup.yml:8
FAILED - RETRYING: Confirm percona pod has been deleted (6 retries left).
FAILED - RETRYING: Confirm percona pod has been deleted (5 retries left).
changed: [localhost -> None] => {"attempts": 3, "changed": true, "cmd": "source ~/.profile; kubectl get pods", "delta": "0:00:00.306519", "end": "2017-11-07 12:48:46.688147", "rc": 0, "start": "2017-11-07 12:48:46.381628", "stderr": "", "stderr_lines": [], "stdout": "NAME                                   READY     STATUS    RESTARTS   AGE\nmaya-apiserver-3416621614-g6tmq        1/1       Running   0          4d\nopenebs-provisioner-4230626287-503dv   1/1       Running   0          4d", "stdout_lines": ["NAME                                   READY     STATUS    RESTARTS   AGE", "maya-apiserver-3416621614-g6tmq        1/1       Running   0          4d", "openebs-provisioner-4230626287-503dv   1/1       Running   0          4d"]}

TASK [set_fact] ********************************************************************************************************
task path: /home/ciuser/openebs/e2e/ansible/playbooks/feature/ha/ha.yml:160
ok: [localhost] => {"ansible_facts": {"flag": "Pass"}, "changed": false}

TASK [Send slack notification] *****************************************************************************************
task path: /home/ciuser/openebs/e2e/ansible/playbooks/feature/ha/ha.yml:168
ok: [localhost] => {"changed": false, "msg": "OK"}
META: ran handlers
META: ran handlers

PLAY RECAP *************************************************************************************************************
localhost                  : ok=25   changed=18   unreachable=0    failed=0
```

File: e2e/ansible/playbooks/feature/ha/ha.yml [hunk 1]
```diff
@@ -0,0 +1,206 @@
...
+       - name: Set pod variables to facts 
+         set_fact:
+           pod_status: "{{ result.stdout.split()[2] }}"
+           pod_restarts: "{{ result.stdout.split()[3] }}" 
+           pod_ip: "{{ result.stdout.split()[5] }}"
+
+       - name: Verify percona is running w/o restarts
+         debug: 
+           msg: "Percona pod health-check is successful"
+         failed_when: "pod_status != 'Running' and pod_restarts != '0'"
```
Comment (AmitKumarDas) [inline e2e/ansible/playbooks/feature/ha/ha.yml]: Can we check through the history of messages or status to confirm ? There might be errors which were temporary but did not result in restarts ? Should we bother those cases ? If at all there were temporary issues can we capture for how long ?

Comment (AmitKumarDas) [reply e2e/ansible/playbooks/feature/ha/ha.yml]: Can there be more verifications/checkpoints not limited to K8s ?

Comment (ksatchit) [reply e2e/ansible/playbooks/feature/ha/ha.yml]: Yes Amit, you are right. We **should** have non-k8s checkpoints to confirm successful HA. One sure-shot way - and the general expectation out of HA is : 

- To ensure no service breakage (no failure of mysql queries  in this case) during HA

For this, I am trying to identify the way to run parallel tasks via ansible - something I have alluded to in the comment about "async module" in the PR.  The plan is to have these added into the test once the method is identified. 

Now we are doing db writes after reschedule only as a post-ha recovery health-check.

**Pull request merged**